esquery.js

```diff
@@ -334,38 +334,63 @@ function subjects(selector, ancestor) {
     return results;
 }
 
+/**
+* @callback TraverseVisitor
+* @param {?external:AST} node
+* @param {?external:AST} parent
+* @param {external:AST[]} ancestry
+*/
+
 /**
  * From a JS AST and a selector AST, collect all JS AST nodes that
  * match the selector.
  * @param {external:AST} ast
  * @param {?SelectorAST} selector
+ * @param {TraverseVisitor} visitor
  * @returns {external:AST[]}
  */
-function match(ast, selector) {
-    const ancestry = [], results = [];
-    if (!selector) { return results; }
+function traverse(ast, selector, visitor) {
+    if (!selector) { return; }
+    const ancestry = [];
     const altSubjects = subjects(selector);
     estraverse.traverse(ast, {
         enter (node, parent) {
             if (parent != null) { ancestry.unshift(parent); }
             if (matches(node, selector, ancestry)) {
                 if (altSubjects.length) {
                     for (let i = 0, l = altSubjects.length; i < l; ++i) {
-                        if (matches(node, altSubjects[i], ancestry)) { results.push(node); }
+                        if (matches(node, altSubjects[i], ancestry)) {
+                            visitor(node, parent, ancestry.slice());
```

Comment:
I had missed this issue yesterday. Let's not copy the ancestry here. Leave it up to the consumers to copy if they want to store it and not have it mutate. The performance hit in the usual case isn't worth the convenience.

Reply:
Ok, I've amended to just pass `ancestry` as is and am only slicing it in the test now.